flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1198,16 +1205,231 @@ class DidNotConverge(Exception):
+        :param current_cluster_state: A callable that returns the current state
+            of the cluster.
+        """
+        dataset = unicode(self._dataset_id)
+
+        # First poll the cluster state to see if there is a new mountpoint that
+        # we should try to use to attempt to write to the dataset.
+        [node] = list(n for n in current_cluster_state().nodes
+                      if n.uuid == self._node_uuid)
+
+        # If this dataset is in the manifestations, then we should be able to
+        # write to the current path provided by flocker.
+        manifestations = node.manifestations or {}
+        write_to_current_path_should_succeed = dataset in manifestations
+
+        paths = node.paths or {}
+        current_path = paths.get(dataset)
+        if write_to_current_path_should_succeed:
+            case.assertIsNotNone(
+                current_path, "Despite being marked as manifested, dataset %s "
+                "did not have a path in the NodeState." % dataset)
+        if current_path:
+            self._known_mountpoints.add(current_path)
+
+        devices = node.devices or {}
+        device = devices.get(self._dataset_id)
+        if device:
+            # Update the device_path that is backing the dataset_id just in
+            # case the volume has changed to a different device path since the
+            # last time we ran.
+            self._device_path = device
+
+        for path in self._known_mountpoints:
+            filename = unicode(uuid4())
+            random_string = unicode(uuid4())
+            path_to_write = path.child(filename)
+            try:
+                path_to_write.setContent(random_string)
+            except OSError:
+                # This indicates that we failed to write to a path provided by
+                # flocker. Assert that this is not the path of a currently
+                # manifest dataset.
+                case.assertFalse(
+                    write_to_current_path_should_succeed and
+                    current_path == path,
+                    "Could not write to path %s despite dataset %s reporting "
+                    "as manifest." % (path_to_write, dataset))
+            else:
+                case.assertTrue(
+                    self._has_file_with_content(filename, random_string),
+                    "Successfully wrote to a path gotten from flocker, "
+                    "but the write was not persisted to the backing block "
+                    "device.")
+
+
+def _empty_node_state(deployer):
+    """
+    Returns an empty node state for the node the deployer is created on.
+    """
+    return NodeState(
+        uuid=deployer.node_uuid,
+        hostname=deployer.hostname,
+        applications=[],
+    )
+
+
+@attributes(["test_case", "deployer", "cluster_state"], apply_with_init=False)
+class BlockDeviceCalculatorTestObjects(object):
+    """
+    Object to house all test objects for the tests in
```

Comment:
Hrm, IMHO, setUp and tearDown are broken for the usual reasons that also back, "prefer composition over inheritance". Especially when I just want to pass a per-test parameter into setUp. In the past I've found these helper objects to be a nice compromise for code re-use.

Obviously this one is a bit more stateful than desired, sorry about that.